Import Cram as Jib CLI

briandealwis opened the pull request:
Uses the `application` plugin to create a runnable distribution.

I'll leave Graal support for a separate PR as the [Palantír Graal plugin](https://github.com/palantir/gradle-graal) doesn't support GraalVM 19.3.x.

briandealwis commented:
What was I thinking 🤦‍♂ Windows uses colons in drive specs

chanseokoh commented on jib-cli/README.md:
````diff
@@ -0,0 +1,56 @@
+
+Use the `application` plugin's `installDist` task to create a runnable installation.
+```sh
+$ ../gradlew installDist
````
We can address this later, but because this requires installing JRE and Gradle, I think we should provide pre-compiled binaries, especially when considering Jib CLI is very general and isn't bound to Java developers.

briandealwis replied:
The distribution doesn't require Gradle, but it does require a JRE (as does Gradle and Maven).  But otherwise 100% agreed: the real end-goal is to use Graal's native-image to produce an executable, but that's not easy to do with the most recent release of Graal (yet).

chanseokoh replied:
I meant by Gradle that to obtain Jib CLI, the doc says to run `./gradlew installDist`. I wonder if we can "pre-compile" `gradlew installDist` on Linux, Mac, and Windows, and distribute them (whatever it is)?

loosebazooka replied:
Ah so we talked about. We can't offer distributed binaries unless we go through an official launch process -- so that will as you mention have to be addressed later.

briandealwis replied:
👍

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,35 @@
+  // for an unpublished library.
+  implementation project(':jib-core').configurations.implementation.dependencies
+  implementation "com.google.guava:guava:${dependencyVersions.GUAVA}"
+  implementation "info.picocli:picocli:${dependencyVersions.PICOCLI}"
```
I'm curious how all these work out as a CLI tool. Does it not require a JRE? Is it a single executable binary that embeds all transitive dependency JARs? Or, does it cache dependencies under `~/.m2`?

briandealwis replied:
Picocli is a nice library for processing command-line arguments.  It also supports more sophisticated command-groups (e.g., `jib build ...` vs `jib jar ...` or `jib inspect ...`), though this code isn't taking advantage of these command-groups yet.

chanseokoh replied:
Hmm... I don't think I got the answers for my questions, but I think I'll just try this branch myself and see what `./gradlew installDist` generates and how exactly it works on my machine.

chanseokoh commented on jib-cli/src/main/java/com/google/cloud/tools/jib/cli/ActualTimestampProvider.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.function.BiFunction;
+
+/** A timetamp provider that uses the actual file last-modification time from the file-system. */
+class ActualTimestampProvider implements BiFunction<Path, AbsoluteUnixPath, Instant> {
```
Is this used somewhere? I can't find it. I'm wary of people using this provider if it's going to be used for creating an image, because it greatly undermines reproducibility of a layer.

briandealwis replied:
It's pulled in from `LayerParserDefinition#parseTimestampsDirective()`.

briandealwis replied:
It's pulled in from `LayerParserDefinition#parseTimestampsDirective()`.

briandealwis replied:
It's pulled in from `LayerParserDefinition#parseTimestampsDirective()`.

chanseokoh replied:
So the question is, should we drop support for `ts=actual`? Jib plugins don't support it (and will never after we build an extension framework). I find think this is very useful, especially when considering the negatives. If someone really wanted to set it to the last modification time of some file, I think they could get easily it in a shell script.

```
              + "  ts=timestamp  set last-modified timestamps:\n"
              + "    actual      use actual values in file-system\n"`
```

However, this doesn't need to to addressed in this PR (in case you agree).

briandealwis commented:
Marked as not-ready as I have a few more tweaks to pull in.

briandealwis commented:
PTAL. Refactored and moved the build functionality into a new `jib build` subcommand.  This should facilitate creating `jib jar` or other subcommands.

chanseokoh commented:
Builds are failing.

briandealwis commented:
Oops, and I hadn't updated to the latest jib-build-plan changes.

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+}
+
+ext {
+    cliMainClass = "com.google.cloud.tools.jib.cli.JibCli"
```
nit: indentation is off, and no double quotes

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+plugins {
+  id "application"
```
nit: 'application' (not double quotes)

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+// use `run` to build and run the app
+// use `installDist` or `distZip` to create an installable application
+application {
+  applicationName = "jib"
```
nit: 'jib` (not double quotes)

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+  testImplementation project(path:':jib-core', configuration:'tests')
+}
+
+sourceSets.test.resources.srcDirs project(':jib-core').sourceSets.test.resources
```
Haven't looked into the test, so let me confirm. Do you need to copy test resources from `jib-core`?

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+  testImplementation "org.mockito:mockito-core:${dependencyVersions.MOCKITO_CORE}"
+  testImplementation "org.slf4j:slf4j-api:${dependencyVersions.SLF4J_API}"
+  testImplementation "com.github.stefanbirkner:system-rules:${dependencyVersions.SYSTEM_RULES}"
+  testImplementation project(path:':jib-core', configuration:'tests')
```
This is probably unnecessary?

chanseokoh commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,36 @@
+}
+
+dependencies {
+  implementation project(':jib-build-plan')
```
Can you see if you can remove this (as we are copying `jib-build-plan` source into `jib-core`)? `jib-plugins-common` does not add this, for example. @loosebazooka may have a better idea.

loosebazooka replied:
Yeah we can remove this I think? But I'm not sure how this build works -- does it just package up the application without depending on anything being in the local maven repo?

briandealwis replied:
I've tried removing this and the build breaks with unresolved symbols 🤷‍♂

loosebazooka replied:
oh yeah, because it's not re-exported. so I think even in jib-core we need to make it an `api` dependency rather than `implementation`?

chanseokoh replied:
So what should be done in this PR?

loosebazooka replied:
I think it's correct for this PR.

loosebazooka commented on jib-cli/build.gradle:
```diff
@@ -0,0 +1,40 @@
+dependencies {
+  implementation project(':jib-build-plan')
+  implementation project(':jib-core')
+  implementation project(':jib-core').configurations.implementation.dependencies
```
We can also remove this line, you would only need this if you're trying to access classes in dependencies of jib-core. We do this in appengine-plugins-common kind of as a hack to not have to re specify the implementation dependencies (that is a little hacky, and maybe we should be more explicit about it -- but that's another issue).
